keps/sig-storage/0033-20181111-transfer-pvc-between-namespaces.md

```diff
@@ -0,0 +1,179 @@
+
+## Summary
+
+Provide a mechanism to allow users to "give" a PVC to a user in another namespace based on mutual acceptance. A transfer should be initiated by updating an existing PVC with a transfer request annotation, and accepted by the intended destination namespace by the recipient requesting a PVC using the standard mechanism and specifying the correct PVC Name and transfer annotation as per the transfer request.
+
+* Provide a Cluster wide mechanism to enable or disable the transfer of PVCs
+* Expose standardized process for transferring a PVC from one Namespace to another
+* Require explicit acceptance of a PVC into another users Namespace, ensuring a malitious user can't just dump PVCs into another users Namespace
+* Utilize existing PVC creation semantics for creating a PVC from a Transfer
+* Honor existing configuration options including quota checks, Storage Class access etc
+
+The process of transferring a PVC/Volume is as follows:
+1. Original user indicates they're willing to ``give`` the volume to another namespace
+2. The receiving user indicates they'd like to ``accept`` the volume into their namespace
+3. When the original user deletes the PVC from their namespace, the claim request from the receiving user is fulfilled
+
+## Motivation
+
+There are a number of uses cases where a user would like to have the ability to transfer an existing PVC from one namespace to another.  This is a valuable workflow for persistent storage and enables the ability to easily duplicate and transfer data sets from one environment to another.  These populated PVCs could be a clone of another volume, a volume from snapshot, or data that was written to the volume via an application (ie a database).
+
+An example use case for this feature would be a cluster segmented into two namespaces; namespace-a for production staging, and namespace-b for production.  There are cases where an application could be developed and tested with the same production data without risking any modification or corruption of data in both environments.  Rather than reproducing the data in both namespaces, it would be much more efficient to be able to clone or restore the data from a snapshot in to a volume and then transfer that new volume to the desired namespace.
+
+### Goals
+
+* Provide a secure and standard mechanism to duplicate data sets on volumes across namespaces in the cluster
+* Enhance existing API's and workflows to alow transferring of volumes across namespaces
+
+### Non-Goals
+
+  While this concept could be extended, volumes are a special case.  Most other resources can just be recreated easily, but this isn't as trivial with large data sets.
+* Introducing a new API
+* Introducing a new Transfer Object
+* Enable transfer for resources other than volumes
+
+## Proposal
+
+
+Update the PVC with a transfer request annotation:
+```yaml
+apiversion: v1
+kind: PersistentVolumeClaim
+metadata:
+    name: pvc-foo
+	annotations:
```

Comment:
This should not be an annotation. It should probably be whole resource.

E.g. 

* Assume you have a PVC object "Foo" in NS1
* Create a PVCTransfer object "Foo" with `sendTo: dev`
* Controller observes this, but waits for a receiver
* Create a PVCTransfer object "Bar" in NS2 with "recvFrom: prod/Foo"
* Controller observes this and does the transfer, deleting both PVCTransfer resources when done

This would have to be seriously thought through and audited for attack vectors.

But why are PVCs special?  What about snashots?  What about other non-storage resources?

Reply:
@thockin thanks for looking at this!  

> This should not be an annotation. It should probably be whole resource.

I started with something along those lines, but I wasn't sure it would be acceptable to add a new "object" like that.  The annotations aren't the best answer here I agree; I was actually hoping to introduce a formal parameter to the PV and PVC objects instead.  I'm happy to look at the other option though if it's preferred.  The flow you describe aligns perfectly by the way.

> This would have to be seriously thought through and audited for attack vectors.

Indeed, one of the reasons I used the pvc and pv objects by themselves was to try and minimize opening up new vectors, it might be worth it though.

> But why are PVCs special? What about snashots? What about other non-storage resources?

Initially I proposed to the sig a generlized transfer resource like you described that could be used for any object, but the more I thought about it it didn't seem to make sense (IOW I talked myself out of it).  The reason being that for things like PODs I didn't think there was a good use case (as opposed to just recreating the POD.  Sure, a POD may have some heavy containers, but recreating data sets (like say a 200Gig DB store) is a bit ugly.  If there's a good use case, or if it's preferred to have consistency across objects on the system I can agree with that.

As far as Snapshots, I have a strong opinion about breaking the linkage between Snapshots and Volumes across Namespaces.  For many back end devices these are linked, and even worse some link their snapshots.  By transferring a Snapshot to a different Namespace that creates some visibility issues for users to those linkages.  Say for example my device uses a cow file or something similar for snapshots, and then each consecutive snapshot is another cow file built from there.  The entire chain is linked; if I transfer one snapshot in the link to another namepsace, the original namespace is now unable to delete any of his/her snapshots or volumes without the new user deleting theirs.  OR, the new user can't delete theirs if there were subsequent snapshots from the originator.

To get around that, it might be good to limit transfer of Snapshots to a flow like:
1. Create Snapshot
2. Create New PVC from Snapshot (now the PVC is it's own independent object)
3. Transfer PVC to new Namespace

That way there's no linkage, and the new user and old user can do anything they normally could without introducing some weird corner cases.